docs/man/sja1105-kmod.1
.TH "sja1105\-kmod" "1" "" "" "SJA1105\-TOOL"
.SH NAME
.PP
\f[I]sja1105.ko\f[] \- Kernel driver module for NXP SJA1105
.SH DESCRIPTION
.PP
\f[I]sja1105.ko\f[] is the kernel driver module (kmod) for the SJA1105
switch chip.
It loads the static switch configuration from user space and uploads it
to the chip, monitors the link status of the port\[aq]s PHYs to
reconfigure the port\[aq]s MAC speed and provides a user space interface
to control the operation of the switch.
.SS STATIC CONFIGURATION
.PP
On driver probe, the kmod loads the static configuration from user space
(default location "\f[I]/lib/firmware/sja1105.bin\f[]") and uploads it
to the chip.
.SS RESET PIN HANDLING
.PP
If defined in device tree, the kmod controls the reset lines of the
switch and PHY chips.
On driver probe, it releses the switch from reset, uploads the static
configuration, then releases the PHYs from reset.
.SS NETWORK DEVICE INTERFACES
.PP
The kmod creates a network interface for each port of the switch to
configure and monitor the PHY connected to this port.
On a link up event, the link speed of the port\[aq]s MAC is adjusted to
match with the PHY\[aq]s speed.
.SS SYSFS USER SPACE INTERFACE
.PP
The kmod exports some sysfs attributes to control and monitor the
switch.
.TP
.B config_upload
Write \[aq]1\[aq] to the file to reload the static configuration from
user space and upload it to the switch chip.
.RS
.RE
.TP
.B device_id
Read the device id of the switch chip.
Reading the file returns the device id, part number and the chip\[aq]s
name.
.RS
.RE
.TP
.B general_status
Read general status information from the switch.
Reading the file returns the status information in text format.
.RS
.RE
.TP
.B port_mapping
Get the mapping between the network interface name and port number for
all ports.
Reading the file returns a "\f[I]netif_name\f[] \f[I]port_number\f[]"
pair for each port.
.RS
.RE
.TP
.B port_status
Read status information for a specific port from the switch.
To select a port, write either the network interface name which is
assigned to the port or the port number to the file prior to reading.
Consecutive reads always return the status of the same port.
.RS
.RE
.TP
.B port_status_clear
Clear the status counters for a specific port.
To clear the counters of a port, write either the network interface name
which is assigned to the port or the port number to the file.
Writing "\f[I]all\f[]" clears the counters of all ports.
.RS
.RE
.TP
.B reg_access
Read or write a register of the switch chip.
\f[B]Register write:\f[] Write the register address and value, separated
by a space character to the file.
\f[B]Register read:\f[] Write the register address to the file prior to
reading.
When reading the file, the register value is read from the switch and
returned in the format "\f[I]reg_address\f[] \f[I]value\f[]".
Consecutive reads always return the same registers value.
.RS
.RE
.SH DEVICE\-TREE
.SS BINDINGS
.TP
.B reset\-gpio
GPIO connected to the reset input of the switch chip.
.RS
.RE
.TP
.B phy\-reset\-gpio
GPIO connected to the reset input of a PHY chip.
Can be specified for each port.
.RS
.RE
.TP
.B phy\-reset\-duration
Length of the reset pule in milliseconds.
.RS
.RE
.TP
.B phy\-reset\-delay
Time before the PHY is ready after reset in milliseconds.
.RS
.RE
.SS DEVICE\-TREE EXAMPLE
.IP
.nf
\f[C]
&dspi0\ {
\ \ \ \ sja1105\@1\ {
\ \ \ \ \ \ \ \ compatible\ =\ "nxp,sja1105t";
\ \ \ \ \ \ \ \ reg\ =\ <0x1>;
\ \ \ \ \ \ \ \ spi\-max\-frequency\ =\ <4000000>;
\ \ \ \ \ \ \ \ #address\-cells\ =\ <1>;
\ \ \ \ \ \ \ \ #size\-cells\ =\ <0>;
\ \ \ \ \ \ \ \ fsl,spi\-cs\-sck\-delay\ =\ <1000>;
\ \ \ \ \ \ \ \ fsl,spi\-sck\-cs\-delay\ =\ <1000>;
\ \ \ \ \ \ \ \ #reset\-gpio\ =\ <&spigpio1_1\ 0\ 1>;
\ \ \ \ \ \ \ \ port\@0\ {
\ \ \ \ \ \ \ \ \ \ \ \ sja1105,port\-label\ =\ "seth5";\ /*\ ETH5\ written\ on\ chassis\ */
\ \ \ \ \ \ \ \ \ \ \ \ phy\-handle\ =\ <&rgmii_phy3>;
\ \ \ \ \ \ \ \ \ \ \ \ phy\-mode\ =\ "rgmii";
\ \ \ \ \ \ \ \ \ \ \ \ #phy\-reset\-gpio\ =\ <&spigpio1_1\ 1\ 1>;
\ \ \ \ \ \ \ \ \ \ \ \ #phy\-reset\-duration\ =\ <1>;
\ \ \ \ \ \ \ \ \ \ \ \ #phy\-reset\-delay\ =\ <120>;
\ \ \ \ \ \ \ \ \ \ \ \ reg\ =\ <0>;
\ \ \ \ \ \ \ \ };
\ \ \ \ \ \ \ \ port\@1\ {
\ \ \ \ \ \ \ \ \ \ \ \ sja1105,port\-label\ =\ "seth2";\ /*\ ETH2\ written\ on\ chassis\ */
\ \ \ \ \ \ \ \ \ \ \ \ phy\-handle\ =\ <&rgmii_phy4>;
\ \ \ \ \ \ \ \ \ \ \ \ phy\-mode\ =\ "rgmii";
\ \ \ \ \ \ \ \ \ \ \ \ reg\ =\ <1>;
\ \ \ \ \ \ \ \ };
\ \ \ \ \ \ \ \ port\@2\ {
\ \ \ \ \ \ \ \ \ \ \ \ sja1105,port\-label\ =\ "seth3";\ /*\ ETH3\ written\ on\ chassis\ */
\ \ \ \ \ \ \ \ \ \ \ \ phy\-handle\ =\ <&rgmii_phy5>;
\ \ \ \ \ \ \ \ \ \ \ \ phy\-mode\ =\ "rgmii";
\ \ \ \ \ \ \ \ \ \ \ \ reg\ =\ <2>;
\ \ \ \ \ \ \ \ };
\ \ \ \ \ \ \ \ port\@3\ {
\ \ \ \ \ \ \ \ \ \ \ \ sja1105,port\-label\ =\ "seth4";\ /*\ ETH4\ written\ on\ chassis\ */
\ \ \ \ \ \ \ \ \ \ \ \ phy\-handle\ =\ <&rgmii_phy6>;
\ \ \ \ \ \ \ \ \ \ \ \ phy\-mode\ =\ "rgmii";
\ \ \ \ \ \ \ \ \ \ \ \ reg\ =\ <3>;
\ \ \ \ \ \ \ \ };
\ \ \ \ \ \ \ \ port\@4\ {
\ \ \ \ \ \ \ \ \ \ \ \ sja1105,port\-label\ =\ "sethi";\ /*\ Internal\ port\ connected\ to\ eth2\ */
\ \ \ \ \ \ \ \ \ \ \ \ phy\-mode\ =\ "rgmii";
\ \ \ \ \ \ \ \ \ \ \ \ reg\ =\ <4>;
\ \ \ \ \ \ \ \ \ \ \ \ fixed\-link\ {
\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ speed\ =\ <1000>;
\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ full\-duplex;
\ \ \ \ \ \ \ \ \ \ \ \ };
\ \ \ \ \ \ \ \ };
\ \ \ \ };
};
\f[]
.fi
.SH BUGS
.SH AUTHOR
.PP
sja1105\-tool was written by Vladimir Oltean <vladimir.oltean@nxp.com>
.SH SEE ALSO
.SH COMMENTS
.PP
This man page was written using pandoc (http://pandoc.org/) by the same
author.


src/kmod/sja1105-kmod.c
/*
 * SPDX-License-Identifier: GPL-2.0
 *
 * Copyright (c) 2018, Sensor-Technik Wiedemann GmbH
 * Copyright (c) 2018, NXP Semiconductors
 */

#define pr_fmt(fmt) KBUILD_MODNAME ": " fmt

#include <linux/module.h>
#include <linux/printk.h>
#include <linux/version.h>
#include <linux/netdevice.h>
#include <linux/spi/spi.h>
#include <linux/errno.h>
#include <linux/firmware.h>
#include <linux/netdev_features.h>
#include <linux/etherdevice.h>
#include <linux/of_mdio.h>
#include <linux/of_net.h>
#include <linux/phy.h>
#include <linux/gpio/consumer.h>
#include "sja1105.h"
#include <lib/include/static-config.h>
#include <lib/include/staging-area.h>
#include <lib/include/dynamic-config.h>
#include <lib/include/spi.h>

MODULE_AUTHOR("Vladimir Oltean <vladimir.oltean@nxp.com>");
MODULE_AUTHOR("Georg Waibel <georg.waibel@sensor-technik.de>");
MODULE_DESCRIPTION("SJA1105 Driver");
MODULE_LICENSE("GPL");

/* Common function used for issuing a reset pulse
 * to either the switch (priv->reset_gpio) or to an individual
 * external PHY (port->reset_gpio).
 */
static int sja1105_hw_reset_chip(struct gpio_desc *gpio,
                                 unsigned int pulse_len,
                                 unsigned int startup_delay)
{
	if (IS_ERR(gpio))
		return PTR_ERR(gpio);

	gpiod_set_value_cansleep(gpio, 1);
	/* Wait for minimum reset pulse length */
	msleep(pulse_len);
	gpiod_set_value_cansleep(gpio, 0);
	/* Wait until chip is ready after reset */
	msleep(startup_delay);

	return 0;
}

/*
 * Load the SJA1105 static configuration from rootfs and unpack it
 * into private data.
 */
int sja1105_load_firmware(struct sja1105_spi_private *priv)
{
	int rc;
	void* buf;
	struct device *dev = &priv->spi_dev->dev;
	const struct firmware *fw_entry;

	/* Load firmware from user space (rootfs) and unpack it */
	rc = request_firmware(&fw_entry, priv->staging_area, dev);
	if (rc) {
		dev_err(dev, "Error requesting firmware file %s\n",
		        priv->staging_area);
		goto err_out;
	}

	/* sja1105_static_config_unpack() wants a non-const data pointer but
	 * fw_entry is const -> copy data to a local buffer */
	buf = kmalloc(fw_entry->size, GFP_KERNEL);
	if (!buf) {
		rc = -ENOMEM;
		goto err_fw;
	}
	memcpy(buf, fw_entry->data, fw_entry->size);

	rc = sja1105_static_config_unpack(buf, fw_entry->size,
	                                  &priv->static_config);
	if (rc < 0) {
		dev_err(dev, "Invalid static configuration data\n");
		goto err_mem;
	}

err_mem:
	kfree(buf);
err_fw:
	release_firmware(fw_entry);
err_out:
	return rc;
}


static void sja1105_cleanup(struct sja1105_spi_private *priv)
{
	struct list_head *pos, *q;
	struct sja1105_port *port;

	sja1105_sysfs_remove(priv);
	sja1105_ptp_clock_unregister(priv);

	list_for_each_safe(pos, q, &(priv->port_list_head.list)) {
		port = list_entry(pos, struct sja1105_port, list);
		if (port) {
			dev_dbg(&priv->spi_dev->dev,
			        "Cleaning up port %s\n",
			        port->net_dev->name);
			if (!IS_ERR(port->reset_gpio))
				gpiod_put(port->reset_gpio);
			if (of_phy_is_fixed_link(port->node))
				of_phy_deregister_fixed_link(port->node);
			else if (port->phy_dev && port->phy_dev->attached_dev)
				phy_disconnect(port->phy_dev);
			sja1105_netdev_remove_port(port);
			list_del(pos);
		}
	}

	if (!IS_ERR(priv->reset_gpio))
		gpiod_put(priv->reset_gpio);

	kfree(priv);
}


static int sja1105_connect_phy(struct sja1105_port *port,
                               const char *port_name)
{
	struct device *dev = &port->spi_dev->dev;
	struct net_device *net_dev = port->net_dev;


	/* Configure reset pin and bring up PHY */
	port->reset_gpio =
#if KERNEL_VERSION(4, 10, 0) <= LINUX_VERSION_CODE
		devm_fwnode_get_gpiod_from_child(dev, "phy-reset",
		                                &port->node->fwnode,
		                                GPIOD_OUT_HIGH,
		                                port->node->name);
#elif KERNEL_VERSION(4, 9, 0) <= LINUX_VERSION_CODE
		devm_get_gpiod_from_child(dev, "phy-reset",
		                          &port->node->fwnode);
#else
		/* TODO there are kernel versions where we are still
		 * incompatible with GPIOD API */
		devm_get_gpiod_from_child(dev, "phy-reset", NULL,
		                          &port->node->fwnode);
#endif
	if (IS_ERR(port->reset_gpio)) {
		dev_dbg(dev, "%s: reset-gpios (PHY reset pin) not defined, ignoring...\n",
		        port_name);
	} else {
		port->reset_duration = 1;  /* preset with default value */
		port->reset_delay = 10;    /* preset with default value */
		of_property_read_u32(port->node, "phy-reset-duration",
		                     &port->reset_duration);
		of_property_read_u32(port->node, "phy-reset-delay",
		                     &port->reset_delay);
		sja1105_hw_reset_chip(port->reset_gpio, port->reset_duration,
		                      port->reset_delay);
	}

	port->phy_node = of_parse_phandle(port->node, "phy-handle", 0);
	if (IS_ERR_OR_NULL(port->phy_node)) {
		if (!of_phy_is_fixed_link(port->node)) {
			dev_err(dev, "%s: phy-handle or fixed-link properties missing!\n",
			        port_name);
			goto err_out;
		}
		/* phy-handle is missing, but fixed-link isn't.
		 * So it's a fixed link.
		 */
		if (of_phy_register_fixed_link(port->node) != 0) {
			dev_err(dev, "Failed to register %s as fixed link!\n",
			        port_name);
			goto err_out;
		}
	} else {
		port->phy_dev = of_phy_connect(net_dev, port->phy_node,
		                               sja1105_netdev_adjust_link,
		                               0 /* flags */, port->phy_mode);
		port->net_dev->phydev = port->phy_dev;
		/* Regardless of error status, decrement refcount now
		 * (of_parse_phandle)
		 */
		of_node_put(port->phy_node);
		if (IS_ERR_OR_NULL(port->phy_dev)) {
			dev_err(dev, "%s: Could not connect to PHY\n",
			        port_name);
			goto err_out;
		}
	}
	return 0;

err_out:
	return -ENODEV;
}

/*
 * This function also performs the firmware request to userspace once
 * it parses the path to the staging area from the DTS. The static
 * configuration is then loaded into the driver private data struct.
 */
static int sja1105_parse_dt(struct sja1105_spi_private *priv)
{
	struct device       *dev = &priv->spi_dev->dev;
	struct device_node  *switch_node = dev->of_node;
	struct device_node  *child;
	const char          *port_name;
	struct sja1105_port *port;
	phy_interface_t     phy_mode;
	u32 port_index;
	u64 speed;
	int rc;

	rc = of_property_read_string(switch_node, "sja1105,staging-area",
	                             &priv->staging_area);
	if (rc) {
		dev_err(dev, "Staging area node not present in device tree!\n");
		goto err_out;
	}

	/* Load firmware from user space (rootfs) and unpack it */
	rc = sja1105_load_firmware(priv);
	if (rc)
		goto err_out;

	for_each_child_of_node(switch_node, child) {
		rc = of_property_read_string(child, "sja1105,port-label",
		                            &port_name);
		if (rc) {
			dev_err(dev, "Invalid port node in device tree\n");
			goto err_out;
		}

		/* Get switch port number from DT */
		rc = of_property_read_u32(child, "reg", &port_index);
		if (rc) {
			dev_err(dev, "Port number not defined in device tree (property \"reg\")\n");
			goto err_out;
		}

		/* Get PHY mode from DT */
		phy_mode = of_get_phy_mode(child);
		if (phy_mode < 0) {
			dev_err(dev, "Failed to read %s phy-mode or phy-interface-type property\n",
			        port_name);
			goto err_out;
		}

		/* Create a net_device for the port */
		port = sja1105_netdev_create_port(priv, port_name);
		if (IS_ERR(port))
			goto err_out;

		port->index = port_index;
		speed = priv->static_config.mac_config[port->index].speed;
		port->auto_speed = (speed == 0) ? 1 : 0;
		port->node = child;
		port->phy_mode = phy_mode;

		/* Connect PHY */
		rc = sja1105_connect_phy(port, port_name);
		if (rc) {
			sja1105_netdev_remove_port(port);
			goto err_out;
		}
		if (IS_ERR_OR_NULL(port->phy_node))
			dev_dbg(dev, "Probed %s as fixed link\n", port_name);
		else
			dev_dbg(dev, "Probed %s: %s\n",
			        port_name, port->phy_dev->drv->name);
		list_add_tail(&(port->list), &(priv->port_list_head.list));
	}

	return 0;

err_out:
	return -ENODEV;
}

static int sja1105_probe(struct spi_device *spi)
{
	int rc;
	struct device *dev = &spi->dev;
	struct sja1105_spi_private *priv;
	const char *chip_name;

	if (!dev->of_node) {
		dev_err(dev, "No DTS bindings for SJA1105 driver\n");
		return -EINVAL;
	}

	priv = kzalloc(sizeof(struct sja1105_spi_private), GFP_KERNEL);
	if (!priv)
		return -ENOMEM;

	INIT_LIST_HEAD(&(priv->port_list_head.list));
	mutex_init(&priv->lock);

	/* Populate our driver private structure (priv) based on
	 * the device tree node that was probed (spi) */
	priv->spi_dev = spi;
	spi_set_drvdata(spi, priv);

	/* Configure the SPI bus */
	spi->mode = SPI_CPHA;
	spi->bits_per_word = 8;
	rc = spi_setup(spi);
	if (rc) {
		dev_err(dev, "Could not init SPI\n");
		goto err_out;
	}

	/* Configure reset pin and bring up switch */
	priv->reset_gpio = devm_gpiod_get(dev, "reset", GPIOD_OUT_HIGH);
	if (IS_ERR(priv->reset_gpio))
		dev_dbg(dev, "reset-gpios (switch reset pin) not defined, ignoring...\n");
	else
		sja1105_hw_reset_chip(priv->reset_gpio, 1, 1);

	/* Probe device */
	rc = sja1105_device_id_get(&priv->spi_setup,
	                           &priv->spi_setup.device_id,
	                           &priv->spi_setup.part_nr);
	if (rc < 0)
		goto err_out;
	chip_name = sja1105_device_id_string_get(priv->spi_setup.device_id,
	                                         priv->spi_setup.part_nr);
	dev_dbg(dev, "Probed switch chip: %s\n", chip_name);

	rc = sja1105_sysfs_init(priv);
	if (rc) {
		dev_err(dev, "Failed to create sysfs entries\n");
		goto err_out;
	}

	/* Parse device tree and bring up PHYs */
	mutex_lock(&priv->lock);
	rc = sja1105_parse_dt(priv);
	mutex_unlock(&priv->lock);
	if (rc < 0)
		goto err_out;

	/* Upload static configuration */
	rc = sja1105_static_config_flush(&priv->spi_setup,
	                                 &priv->static_config);
	if (rc < 0)
		goto err_out;
	dev_dbg(dev, "Uploaded static configuration to device\n");

	rc = sja1105_ptp_clock_register(priv);
	if (rc < 0)
		goto err_out;

	return 0;
err_out:
	sja1105_cleanup(priv);
	return rc;
}

static int sja1105_remove(struct spi_device *spi)
{
	struct sja1105_spi_private *priv = spi_get_drvdata(spi);

	sja1105_cleanup(priv);
	return 0;
}

static const struct of_device_id sja1105_dt_ids[] = {
	{ .compatible = "nxp,sja1105e" },
	{ .compatible = "nxp,sja1105t" },
	{ .compatible = "nxp,sja1105p" },
	{ .compatible = "nxp,sja1105q" },
	{ .compatible = "nxp,sja1105r" },
	{ .compatible = "nxp,sja1105s" },
	{ /* sentinel */ },
};
MODULE_DEVICE_TABLE(of, sja1105_dt_ids);

static struct spi_driver sja1105_driver = {
	.driver = {
		.name  = "sja1105",
		.owner = THIS_MODULE,
		.of_match_table = of_match_ptr(sja1105_dt_ids),
	},
	.probe    = sja1105_probe,
	.remove   = sja1105_remove,
};

static int __init sja1105_init(void)
{
	int rc;

	rc = spi_register_driver(&sja1105_driver);
	if (rc < 0)
		printk(KERN_ERR "NXP SJA1105 driver failed to register\n");

	return rc;
}
module_init(sja1105_init);

static void __exit sja1105_exit(void)
{
	spi_unregister_driver(&sja1105_driver);
}
module_exit(sja1105_exit);

